Add semantic conventions for process resource (#635)
open-telemetry/opentelemetry-specification: specification/resource/semantic_conventions/README.md

[ARN]:https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
[FunctionDirectory]: https://github.com/Azure/azure-functions-host/wiki/Retrieving-information-about-the-currently-running-function

> ### Process

Reviewer: Please don't forget adding it to the TOC above (we use `markdown-toc` in some of the other .md files for that).

Author: Updated - thnx!

**type:** `process`

**Description:** An operating system process.

> | Attribute  | Description  | Example  |

Reviewer 2: What about adding a `Required` column? (I just realized some sections in the current document don't have such column, but we should include them, for clarity purposes).

|---|---|---|
| process.pid | Process identifier (PID). | `1234` |

> | process.name | Process name. | `OpenTelemetry Collector` |

Author: Removed/replace (see comment below)

> | process.command | The process executable. | `cmd/otelcol` |

Author: Combining this feedback with the comment on "process.name" above, I've removed "process.command" and changed the definition of "process.name" to a [third alternative](https://stackoverflow.com/a/14176059): "Last component of the path to the executable."

This matches what is stored in the Name field in [/proc/[pid]/status](https://man7.org/linux/man-pages/man5/proc.5.html) (although the proc docs just refer to this as "Command run by this process") and appears to match how Windows derives the process name used in Windows task manager.

My reasoning being that the first element of the command is largely redundant since we provide the full commandline as a separate attribute. Also I believe this matches what is shown in commands like `top` (correct me if I'm wrong)

> | process.command | The process executable. | `cmd/otelcol` |

Reviewer 3: The current conventions are still unclear on whether the "invoked path" or the resolved path should be used. Reading between the lines, at least the "initial full path" (don't known the proper name for that) path (e.g. /bin/ls instead of ls) should be used, but first that should also be stated more explicitly and second its still unclear what should happen if that initial resolved path is a symbolic link, as is common on Unix/Linux.

> | process.command | The process executable. | `cmd/otelcol` |

Author: I think you were interpreting it differently than I intended. From my understanding, there's four things we could include here, and I guess they could all be relevant, so I've updated the PR to include all four of them as options.

The descriptions may still not be clear enough - let me know if they make sense. Here's a more detailed explanation of what I intended:

1. Name: the name (not full path) of the "resolved" executable (i.e. the `Name` in `/proc/[pid]/status`)
2. Exe: the full path of the "resolved" executable (i.e. the target of the `/proc/[pid]/exe` symlink)
3. Command: the name of the "invoked" executable (i.e. the first null terminated string in `/proc/PID/cmdline`)
4. Command Line: the full command used to "invoke" the process (i.e. all strings in `/proc/PID/cmdline` concatenated with a space)

#1,2,3 correspond to the first three bullet points here: https://askubuntu.com/a/21698

> | process.commandline | The full command used to launch the process. | `cmd/otecol --config=config.yaml` |

Reviewer 3: Should this be a single string or a list of strings? In some environments one is easier to obtain than the other.

Author: [/proc/[pid]/cmdline](https://man7.org/linux/man-pages//man5/procfs.5.html) returns a list of strings separated by null characters so I guess its slightly easier to obtain as a list of strings, but requires some parsing either way.

In Windows, you just get back a single [CommandLine](https://docs.microsoft.com/en-us/windows/win32/api/winternl/ns-winternl-rtl_user_process_parameters#members) string.

Overall, I think a single string should be fine here.

Reviewer 3: So should the string be separated by spaces or NULs or is that system dependent? I think we should be more specific here. I would recommend specifying that one has both options: A single string if that is the natural representation, or a list otherwise.

As you said, on Windows, a single string is the natural representation (it is up to programs to split the string a whitespace or whatever crazy algorithm they want to use; usually this is done by the C runtime, but even of these I'm sure multiple dialects exist) and on Unix a list would be more natural (the shell splits the argument list and the program receives an actual list). Though in some frameworks/languages it might be easier to get one or the other representation regardless of the OS-native one.

I also recommend explicitly giving examples like "Should be set to the value of GetCommandLineW" on Windows.

Author: Yea that's a good idea. I've updated with examples of where to get the values from on Unix/Windows.

[ARN]:https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
[FunctionDirectory]: https://github.com/Azure/azure-functions-host/wiki/Retrieving-information-about-the-currently-running-function

### Process

**type:** `process`

**Description:** An operating system process.

| Attribute  | Description  | Example  | Required |
|---|---|---|--|
| process.pid | Process identifier (PID). | `1234` | Yes |

> | process.name | Last component of the path to the executable. | `otelcol` | Yes |

Reviewer: Have you considered the full path to the executable? This could hold valuable information, especially if you have executables with the same name but in different foldes (e.g. one subfolder per version or tenant).
How about adding two attributes, `process.executable_path` for the full path and `process.executable_name`, with requiring either one of them? You could recommend providing the full path then or even require it, unless only the name is available.

Author: This corresponds to `process.exe` and `process.name` respectively in the updated spec. See the discussion above for more details.

[ARN]:https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
[FunctionDirectory]: https://github.com/Azure/azure-functions-host/wiki/Retrieving-information-about-the-currently-running-function

### Process

**type:** `process`

**Description:** An operating system process.

| Attribute  | Description  | Example  | Required |
|---|---|---|--|
| process.pid | Process identifier (PID). | `1234` | Yes |

> | process.name | The name of the process executable. On Unix based systems, can be set to the Name in `proc/[pid]/status`. On Windows, can be set to the base name of `GetProcessImageFileNameW`. | `otelcol` | See below |

Reviewer: ```suggestion
| process.executable_name | The name of the process executable. On Unix based systems, can be set to the Name in `proc/[pid]/status`. On Windows, can be set to the base name of `GetProcessImageFileNameW`. | `otelcol` | See below |
```

> | process.exe | The full path to the process executable. On Unix based systems, can be set to the target of `proc/[pid]/exe`. On Windows, can be set to `GetProcessImageFileNameW`. | `cmd/otelcol` | See below |

Reviewer: ```suggestion
| process.executable_path | The full path to the process executable. On Unix based systems, can be set to the target of `proc/[pid]/exe`. On Windows, can be set to `GetProcessImageFileNameW`. | `cmd/otelcol` | See below |
```

`process.exe` reads like an example for an executable on Windows IMHO.

I'd also suggest renaming `process.name` to `process.executable_name` to make them both unambiguous, consistent and self-explanatory.

Author: Updated. I was a little 50/50 on this change, as the new keys seem a little verbose but I guess clarity is more important than succinctness.

Reviewer: @james-bebbington Yeah opinions are often split on this 😄
I'm usually pro clarity. Given the typical length of the whole executable path or command line, the attribute keys are negligible from a performance or memory overhead perspective.

> | process.executable_path | The full path to the process executable. On Unix based systems, can be set to the target of `proc/[pid]/exe`. On Windows, can be set to the result of `GetProcessImageFileNameW`. | `cmd/otelcol` | See below |

Reviewer 4: Should this be an absolute path in the example?

Author: Updated the example path. But not sure what you mean by diff selection?

Reviewer 4: I just meant I wasn't sure if it was clear based on my original comment which one I was referring to since github shows a range of lines.

> | process.executable_name | The name of the process executable. On Unix based systems, can be set to the `Name` in `proc/[pid]/status`. On Windows, can be set to the base name of `GetProcessImageFileNameW`. | `otelcol` | See below |
> | process.executable_path | The full path to the process executable. On Unix based systems, can be set to the target of `proc/[pid]/exe`. On Windows, can be set to the result of `GetProcessImageFileNameW`. | `/usr/bin/cmd/otelcol` | See below |

Reviewer 5: Should `executable` be a hierarchy level?

```suggestion
| process.executable.name | The name of the process executable. On Unix based systems, can be set to the `Name` in `proc/[pid]/status`. On Windows, can be set to the base name of `GetProcessImageFileNameW`. | `otelcol` | See below |
| process.executable.path | The full path to the process executable. On Unix based systems, can be set to the target of `proc/[pid]/exe`. On Windows, can be set to the result of `GetProcessImageFileNameW`. | `/usr/bin/cmd/otelcol` | See below |
```

| process.command | The command used to launch the process (i.e. the command name). On Unix based systems, can be set to the zeroth string in `proc/[pid]/cmdline`. On Windows, can be set to the first parameter extracted from `GetCommandLineW`. | `cmd/otelcol` | See below |

> | process.commandline | The full command used to launch the process. This can be a nil- or space-delimited string. On Unix based systems, can be set to the full `proc/[pid]/cmdline` string. On Windows, can be set to the result of `GetCommandLineW`. | `cmd/otecol --config=config.yaml` | See below |

Reviewer 3: I don't think using "nil" is clear here. Also, this is ambiguous. If I have only a single string with spaces but no nil, it is impossible to determine if I have `some\ command\ with/spacesinpath` and no arguments, or a GetCommandLineW-encoded command line.

Author: There is also concern around using two different data types to represent the same attribute value: https://github.com/open-telemetry/opentelemetry-specification/pull/635#discussion_r437573287. It might create additional complexity on back end systems to have to manage potentially different data types for the same attribute, though I'm not sure how much of a concern this would be? I guess parsing null-characters on the back end would likely be equivalently complex.

Note this isn't strictly ambiguous btw. If that was a command with no arguments, and you were looking at the Unix representation, it would have a null-character as the last character in the string. On Windows it would not. Admittedly that is not a great representational difference, and could be difficult to distinguish in some programming languages.

After thinking about this some more, I think I do prefer the list of strings approach, but I'm not sure I'm aware of all the pros/cons. For now I have changed it back to that, and updated the example for each OS accordingly. But for the record here's the three options considered and pros/cons:

1. Join arguments with a string:
  **Advantages:** Most consistent representation across OSs.
  **Disadvantages:** Potentially lossy/non-trivial conversion on Unix - do we need to enquote parameters with spaces in them (and escape internal quotes), etc?
2. Pass through string returned from OS as is (null-delimited on Unix):
  **Advantages:** No processing needed by code producing telemetry, can just provide what OS returns.
  **Disadvantages:** Back end needs to parse null-characters.
3. Convert to list of strings (for Unix only): 
  **Advantages:** Clean ("properly typed") representation.
  **Disadvantages:** Inconsistent data type for the same attribute between OSs, and back end needs to handle this differently.

FYI @arminru note I've reverted the changes based on your last comment since you approved. Please take another look.

@igorpeshansky you may also be interested.

Reviewer 4: Interestingly ps on Mac and Linux doesn't appear to apply any quoting so if there are spaces or quotes in in your argument (e.g. `app --set "foo bar"`) ps doesn't appear to quote them. We could quote them (https://godoc.org/github.com/gonuts/go-shellquote#Join [1]). I can't think of any downsides in particular when it comes to filtering on quoted values. I think in most cases you're not going to be doing filtering on values that have quotes in them or other escaped characters. I imagine the common case is you're searching for commands that have basic things like "repl-factor" or some such in it.

[1] Not endorsing it, just searched for a shell escape library in go.

> | process.commandline | The full command used to launch the process. This can be a list of strings (ordered arguments), or a single string with arguments separated by a space. On Unix based systems, can be set to the list of strings returned by `proc/[pid]/cmdline`. On Windows, can be set to `GetCommandLineW`. | `cmd/otecol --config=config.yaml` | See below |

Reviewer: I don't think we ever defined an attribute to be either string _or_ array of string, they all have a single, well-defined type so far if I'm not mistaking.
How about making it a single, `\n`-delimited string instead of a list in the former case?

Also, _"On Windows, can be set to `GetCommandLineW`"_ should be changed to _"On Windows, this can be set to the return value of `GetCommandLineW`"_ to avoid misunderstandings.

Author: I was going off the suggestion above - https://github.com/open-telemetry/opentelemetry-specification/pull/635#discussion_r436758753. I've changed this to "This can be a nil- or space-delimited string" which hopefully captures the intent more explicitly.

> | process.commandline | The full command used to launch the process. The value can be either a list of strings representing the ordered list of arguments, or a single string representing the full command. On Unix based systems, can be set to the list of null-delimited strings extracted from `proc/[pid]/cmdline`. On Windows, can be set to the result of `GetCommandLineW`. | Unix: `[ cmd/otecol, --config=config.yaml ]`, Windows: `cmd/otecol --config=config.yaml` | See below |

Reviewer 5: I usually see `commandline` as two words. Do we want to build naming hierarchy here as well? Or, just insert a separator?

Author: Changed this to `process.command_line` for now. I'm not sure if `process.command.name` would be a reasonable name for the "command"

Reviewer 5: SGTM :+1:

| process.owner | The username of the user that owns the process. | `root` | No |

> At least one of `process.name`, `process.exe`, `process.command`, or `process.commandline` is required.

Reviewer: ```suggestion
At least one of `process.executable_name`, `process.executable_path`, `process.command`, or `process.commandline` is required.
```
